Admin api companion

taksan opened the pull request:
This PR will introduce the ability to provide companions from the Admin API server

moufmouf commented:
Ooooh! Pretty nice!
I see this is WIP but being able to read the companion list from admin is definitely a must have so I'm eager to merge this one!

taksan commented:
Yeah, the feature is working now (from my local tests) :). I just need to figure out how to write e2e tests for this and then I'll clear the WIP flag

taksan commented on pusher/tests/AuthenticatedProviderControllerTest.ts:
```diff
@@ -0,0 +1,139 @@
+
+class FakeRequest {
+    params: object = {}
+    constructor(public path_query: string = "", private headers: object = {"Authorization":"token"}) {
```
Not a real secret, this is just fake data for test

taksan commented:
@moufmouf ready for review

nolway commented on pusher/src/Services/LocalCompanionSevice.ts:
```diff
@@ -1,14 +1,20 @@
+import { CompanionList } from "../Messages/JsonMessages/CompanionTextures";
 
-export const COMPANION_RESOURCES: CompanionResourceDescriptionInterface[] = [
+const defaultCompanionList: CompanionList = [
```
I think the companion list must be a json file into the pusher data folder like wokas.

nolway commented on messages/JsonMessages/CompanionTextures.ts:
```diff
@@ -0,0 +1,24 @@
+        description: "The URL of the image of the texture.",
+        example: "http://example.com/resources/characters/pipoya/Cat 01-1.png",
+    }),
+    behaviour: extendApi(z.string(), { description: "The companion behavior (dog|cat)", example: "dog" }),
```
An enum check was preferable here to a string check to avoid unknown string value. I have also add a simple behavior if we want implement another thing who isn't a dog or a cat.
```ts
behaviour: extendApi(z.enum(['simple', 'dog', 'cat']), { description: "The companion behavior (simple|dog|cat)", example: "dog" }),
```

taksan replied:
Done

taksan commented:
@nolway Requested changes applied.

taksan commented:
@nolway Any other comments on this PR?

moufmouf commented on messages/JsonMessages/CompanionTextures.ts:
```diff
@@ -0,0 +1,27 @@
+
+//The list of all the companion textures
+
+export const companionTexture = z.object({
+    name: extendApi(z.string(), { description: "The name of the texture.", example: "dog1" }),
+    img: extendApi(z.string(), {
+        description: "The URL of the image of the texture.",
+        example: "http://example.com/resources/characters/pipoya/Cat 01-1.png",
+    }),
+    behaviour: extendApi(z.enum(["simple", "dog", "cat"]), {
+        description: "The companion behavior (simple|dog|cat)",
+        example: "dog",
+    }),
+});
```
Hey @taksan,

I have several questions/remarks on this.

The "behaviour" parameter seems interesting but is currently unused. Do you plan anything special afterwards? (if yes, I'm curious to know what ;), if no, we can probably remove this)

Also, as you probably noticed, we refactored the "select your woka" and "customize your woka" scenes, but we did not refactor the "companion" scene. We will do this in the future, obviously.

Right now, the list of wokas looks something like this:

```json
[
      {
        "name": "Default Collection",
        "textures": [
          {
            "id": "94f39337-4da3-47b5-b8e1-b9528aa2698a",
            "name": "Accessory 1",
            "url": "https://play.staging.workadventu.re/resources/customisation/character_accessories/character_accessories1.png"
          },
          {
            "id": "8453d808-2e39-48a4-abdb-c91b0bf2b4e0",
            "name": "Accessory 2",
            "url": "https://play.staging.workadventu.re/resources/customisation/character_accessories/character_accessories2.png"
          },
          {
            "id": "ba860312-0dfc-4786-b911-499cf4d0b4eb",
            "name": "Accessory 23",
            "url": "https://play.staging.workadventu.re/resources/customisation/character_accessories/character_accessories23.png"
          },
        ]
      }
    ]
```

In the new Woka scene, Woka can belong to several "collections". We will probably use the same "collections" mechanism for companions (we could categorize companions as "Cats", "Dogs", "Robots", "Pokemons" or even "NFTs" for instance.

So even if the companions list does not **yet** support the notion of categories, it might be a good idea to add those right now. This will avoid having to change the Admin API signature in a future release.

moufmouf commented:
Another thing I'm thinking about:

With this PR, you are adding a `ADMIN_API_FEATURES` environment variable.
I understand the reason for this: since we don't have necessarily support for the companion endpoint on our admin, this is a way to improve an admin gradually. This is actually a nice idea.

However, this puts an additional burden on the person installing WorkAdventure that will now have to keep the `ADMIN_API_FEATURES` environment variable in sync with the admin he uses. I'm afraid this could be the source of nasty bugs that would be quite hard to pinpoint.

Here is a possible alternative:

What if we added a new "/capabilities" endpoint to the Admin API, that would return the list of supported endpoints (possibly with a version number).

Something like:

```
GET /capabilities

{
    "api/companion/list": 1,
    "api/woka/list": 1,
    "api/room/access": 1,
   ...
}
```

On pusher start, we could query this endpoint and use the admin accordingly. And we would not need the `ADMIN_API_FEATURES` environment variable at all.

What do you think?

taksan commented:
@moufmouf implementation is complete, but there's still the conflict with develop branch. dev branch is broken as of now, so I'll wait until it's health again beforing fixing the conflitcts

taksan commented:
Ok, develop broken was my mistake for not updating docker-compose. It's ok now

taksan commented:
@moufmouf this is ready for re-review.

taksan commented:
@moufmouf  I believe this is ready for a review round. I fixed the tests and refactored the whole thing to adapt to the changes that have been performed by the front+pusher merge.

To make it easier to test, I updated my sample admin api repository to implement the new required endpoints:

https://github.com/taksan/workadventure-sample-admin-api/

So, that would be a place to start any manual tests.

And to complete, here's a screenshot of a new companion :)

![image](https://user-images.githubusercontent.com/96873/200722511-1b23da85-e2ca-4c85-8410-3e46ed1e0942.png)

taksan commented:
@moufmouf I addressed all the review points, including adding a capabilities api, so it's probably ready to go :)

taksan commented:
@moufmouf rebased, fixed merge conflicts and ajusted imports to comply with new message package.

Again, it's good to go, I believe.

moufmouf commented:
So cool!
I started working on the conflicts at #2815 but had no time to finish. Thanks a lot for all the work.

Let's merge this! :+1:
